source/fundamentals/monitoring.txt

```diff
@@ -0,0 +1,406 @@
+
+Connection Pool Events
+~~~~~~~~~~~~~~~~~~~~~~
+
+Connection pool events are events related to a **connection pool** held by the driver.
+A connection pool is a set of open TCP connections your driver maintains with
+a MongoDB instance. Connection pools help reduce the number of network handshakes
+your application needs to perform with a MongoDB instance, and can help your
+application run faster.
+
+To monitor connection pool events, write a class that implements the
+``ConnectionPoolListener`` interface and register an instance of that class with your
+``MongoClient`` instance.
+
+Example
+^^^^^^^
+
+This example shows how to make a listener class that prints a message each time
+you check out a connection from your connection pool.
+
+The following code defines the ``ConnectionPoolLibrarian`` class which implements the
+``ConnectionPoolListener`` interface.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
+   :language: java
+   :dedent:
+   :start-after: start cp-listener-impl
+   :end-before: end cp-listener-impl
+
+The following code adds an instance of the ``ConnectionPoolLibrarian`` class to a
+``MongoClient`` object. The code then runs a database command to test the
+librarian.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
+   :language: java
+   :dedent:
+   :start-after: start monitor-cp-example
+   :end-before: end monitor-cp-example
+
+The above code snippet should produce output that looks like this:
+
+.. code-block:: none
+   :copyable: false
+
+   Let me get you the connection with id 21...
+   Let me get you the connection with id 21...
+
+It looks like our find operation requested two connections from the pool.
+If you are curious why the driver uses two connections in one find operation, 
+`you can find the answer in this line of the MongoDB driver source code <https://github.com/mongodb/mongo-java-driver/blob/a82ce71d0aa7da9c6455f1b52ae24ed3cd4c1f62/driver-sync/src/main/com/mongodb/client/internal/FindIterableImpl.java#L203>`__.
+
+For more information on the classes and methods mentioned in this section, see
+the following API documentation:
+
+- :java-docs:`ConnectionPoolListener <apidocs/mongodb-driver-core/com/mongodb/event/ConnectionPoolListener.html>`
+- :java-docs:`MongoClientSettings <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html>`
+- :java-docs:`MongoClient <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html>`
+- :java-docs:`ConnectionCheckedOutEvent <apidocs/mongodb-driver-core/com/mongodb/event/ConnectionCheckedOutEvent.html>`
+- :java-docs:`ConnectionCheckOutFailedEvent <apidocs/mongodb-driver-core/com/mongodb/event/ConnectionCheckOutFailedEvent.html>`
+
+.. _monitoring-jmx:
+
+Monitor Connection Pools with JMX
+---------------------------------
+
+You can monitor connection pool events using **Java Management Extensions (JMX)**. 
+JMX provides tools to monitor applications and devices.
```

Comment:
Issue: This sentence leaves me wondering what tools? And in what type of situation is it useful?

Reply:
I feel it is appropriate to provide an unsatisfactory definition of JMX here as it is not the intention of this guide to explain JMX to our readers. The goal of this definition is to give readers enough context to consult Oracle documentation on JMX.